[v1.5][T09-B1] EduBuddy

kengsengg opened the pull request:
@chuakunhong @johnnychanjx @TeyXinHui

dalessr commented on docs/DeveloperGuide.adoc:
```diff
@@ -800,18 +805,24 @@ Priorities: High (must have) - `* * \*`, Medium (nice to have) - `* \*`, Low (un
 |`* * *` |new user |see usage instructions |refer to instructions when I forget how to use the App
 
-|`* * *` |user |add a new person |
+|`* * *` |teacher with many students |find a student by name |retrieve details of a student without having to go through the entire list
```
No need to repeat the features the current version of addressbook already has. Try to think of some features closely related to your target users.

dalessr commented on docs/DeveloperGuide.adoc:
```diff
@@ -844,15 +855,107 @@ Use case ends.
+* 2a. The list is empty.
++
+Use case ends.
+
```
If the contact list is empty, it actually shows an empty-list message to user.

dalessr commented on docs/DeveloperGuide.adoc:
```diff
@@ -844,15 +855,107 @@ Use case ends.
+1.  User requests to list or find persons
+2.  AddressBook shows a list of persons
+3.  User requests to add an appointment for a specific person in the list
+4.  User inputs the selected date and time for the appointment
```
This does not follow the use case format. Use case should be interaction between user and system, thus two user actions cannot appear continuously.

dalessr reviewed (commented):
@kengsengg @TeyXinHui @chuakunhong @johnnychanjx Some comments added. Can make changes accordingly.

kengsengg commented:
**Enhancements**
Kun Hong:  Morph Phone input into NRIC input to reflect EduBuddy requirements
Keng Seng: Add Sort Command to sort students by name in alphabetical order
Johnny:       Change case sensitivity for command word for standardized input
Xin Hui:       Add Delete Tag Command to delete specified tag from every student

dalessr commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -34,6 +34,10 @@ public LogicManager(Model model) {
+
+
+
+
```
It is ok to use single empty line to separate the code. This is not align to the rest of the code.

dalessr commented on src/main/java/seedu/address/logic/parser/TagDeleteCommandParser.java:
```diff
@@ -0,0 +1,32 @@
+     * and returns an TagDeleteCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
+     */
+    public TagDeleteCommand parse(String args) throws ParseException {
```
What if there is space character in the tag? If the space happens to be before/after the tag, you might want to trim the argument before adding it to Tag.

dalessr commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -147,12 +159,53 @@ public boolean removePerson(Person key) throws PersonNotFoundException {
+    /**
+     * Removes a specific tag from an individual person and updates the person's information.
+     * @param tag
+     * @param person
```
Description for the parameters are needed, although it is obvious. This is because other developers may not know the link between the tag and person here

dalessr commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -147,12 +159,53 @@ public boolean removePerson(Person key) throws PersonNotFoundException {
+        Set<Tag> tagList = new HashSet<>(person.getTags());
+        if (tagList.remove(tag)) {
+            Person newPerson = new Person(person.getName(), person.getNric(),
+                    tagList);
```
This line is actually short enough. Why needs a new line?

dalessr commented on src/main/java/seedu/address/ui/PersonCard.java:
```diff
@@ -13,6 +13,8 @@
 
     private static final String FXML = "PersonListCard.fxml";
+    private static final String[] tagColors = { "red" , "orange", "yellow", "green", "blue", "violet", "indigo",
+                                                "purple" , "pink" , "grey"};
```
Just a suggestion: try to choose the tag colors which users can see the word inside clearly :)

dalessr reviewed (commented):
@kengsengg @TeyXinHui @johnnychanjx @chuakunhong Some comments added.

Please note that:
- The target user needs to be specified in the user guide/developer guide. I noticed that you add it in the readme file. It is good, but the users won't look at the readme file! :)
- Update your user guide/developer guide to show the new features added

dalessr commented:
@kengsengg @TeyXinHui @johnnychanjx @chuakunhong 
There is one thing I forgot to mention:

- **Think about the major and minor features you want to implement, and update your user guide to show all the new features (commands), even though you haven't started yet!**

kengsengg commented:
**v1.2 Enhancements**
Kun Hong: Add TagReplace Command that allows a specified tag to be replaced by an intended tag
Keng Seng: Enable Sort Command to sort by tags as well 
Johnny: Redesign the browser panel to fit our user target profile
Xin Hui: Add new Subject class which records subjects taken by students and their grades

dalessr commented:
@kengsengg @TeyXinHui @johnnychanjx @chuakunhong Good job 👍 

Here are something you need to take note:
- I have seen some `coming in v2.0` labels inside user guide, remember to put in the enhancements you have done and label them as `since v1.x`. All the other original features can also be labeled as `since v1.0` to make them consistent
- Inside use guide, the detailed command and example of the new features can be included now. The key point is that you need to come up with the design before implementation
- User guide can include screenshots about how to use the command and what is the expected outcome
- Update `FAQ` and `Command Summary` in user guide accordingly
- Add more label inside the issue tracker (i.e. priority, status)
- For the issues labeled as `story`, the benefit part (after `so that`) should be put inside the description of this issue instead of title
- Make sure all the PRs are linked to the corresponding issues using the issue number
- Write your project portfolio

kengsengg commented:
v1.3 Enhancements
Kun Hong: Add Remarks Command which allows teacher to add in additional remarks 
Keng Seng: Add Appointment Command which allows teacher to set an appointment with student
Johnny: Able to create a student profile page in the browser panel after adding the student
Xin Hui: Able to view the calculated L1R5 score of the selected student

kengsengg commented:
**v1.4 Enhancements**
Kun Hong: Implement AddRemark and DeleteRemark Command 
Keng Seng: Enable the list of appointments to be viewed on the calendar UI
Johnny: Reflect updated student's name, IC, subjects on the student profile page
Xin Hui: Improve the input command for subjects

chuakunhong committed:
Managed to get the command parser to recognise the nextofkin command.

TeyXinHui committed:
revert back SelectCommand

TeyXinHui committed:
Create a new StreamCommand

TeyXinHui committed:
update and add new tests

TeyXinHui committed:
update userguide

TeyXinHui committed:
fix checkstyle

TeyXinHui committed:
Merge pull request #13 from CS2103JAN2018-T09-B1/Week12Testing

Week12 testing update

TeyXinHui committed:
update person test case due to new person attributes

TeyXinHui committed:
fix checkstyle

TeyXinHui committed:
fix more checkstyle

TeyXinHui committed:
Merge pull request #183 from TeyXinHui/master

revert back select command, add new StreamCommand and tests, update userguide

johnnychanjx committed:
Merge pull request #18 from CS2103JAN2018-T09-B1/Week12Testing

Week12 testing

johnnychanjx committed:
Added editing page

johnnychanjx committed:
Add more editing support

johnnychanjx committed:
Added many things

1) Added support for many commands
2) Allows updating of current HTML file after editing information, still requires restart
3) Added calendar to student profile page instead
4) Created a default page.
5) changed logo and name of file

so much done!!!

johnnychanjx committed:
Fixed Checkstyle

Fixed checkstyle errors

johnnychanjx committed:
Merge pull request #184 from johnnychanjx/Week12

Added a bunch of things, ref comments

TeyXinHui committed:
Merge pull request #14 from CS2103JAN2018-T09-B1/Week12Testing

Week12 testing update again

chuakunhong committed:
This is to add the cca.

chuakunhong committed:
Tried next of kin...

chuakunhong committed:
Clear checkstyle

kengsengg committed:
Created new classes for Appointment panel

kengsengg committed:
Added fxml files for AppointmentList

kengsengg committed:
MainWindow: added appointmentListPanel

kengsengg committed:
Implemented getFilteredAppointmentList() method

kengsengg committed:
MainWindow.fxml: created appointment panel on UI and adjusted settings

kengsengg committed:
Fixed tests

kengsengg committed:
Added XmlAdaptedAppointment

kengsengg committed:
Added updateFilteredAppointmentList() method

kengsengg committed:
Saved AppointmentList data into file and show on UI

chuakunhong committed:
Cca has name and position, pass checkstyle and pass testcases.

chuakunhong committed:
Merge pull request #21 from CS2103JAN2018-T09-B1/pr/185

Pr/185

chuakunhong committed:
Merge pull request #22 from CS2103JAN2018-T09-B1/Week12Testing

Week12 testing

chuakunhong committed:
Merge branch 'Testing' into Week12

# Conflicts:
#	src/main/java/seedu/address/logic/commands/AddCommand.java
#	src/main/resources/StudentPage/James.html
#	src/main/resources/StudentPage/John.html

chuakunhong committed:
Can pass test!

chuakunhong committed:
Removed the element, cca position.

TeyXinHui committed:
remove invalid subject combination exception

TeyXinHui committed:
forgot to stage these

chuakunhong committed:
Trying to pass travis.

kengsengg committed:
Edited classes to include info attribute

johnnychanjx committed:
UI Support for Jar creation

TeyXinHui committed:
fix checkstyle

TeyXinHui committed:
Merge pull request #186 from TeyXinHui/master

removed invalid subject exception

chuakunhong committed:
Merge branch 'Week12Testing' into Week12

chuakunhong committed:
Please pass travis...

johnnychanjx committed:
Merge pull request #188 from johnnychanjx/Week12

UI Support for Jar creation

chuakunhong committed:
Merge branch 'Week12' of https://github.com/chuakunhong/main into Week12

chuakunhong committed:
Merge branch 'Week12Testing' into Week12

chuakunhong committed:
Please pass Travis...

chuakunhong committed:
Please go away trailing white spaces..

chuakunhong committed:
Merge pull request #185 from chuakunhong/Week12

Got cca with name and position plus NextOfKin has phone number and email.

kengsengg committed:
Fixed checkstyle and failed tests

kengsengg committed:
SampleDataUtil: Added sample appointment data

kengsengg committed:
AddressBookChangedEvent: included appointments

kengsengg committed:
Collated files

chuakunhong committed:
Updated the UserGuide on commands for InjuriesHistory and the Cca.

TeyXinHui committed:
Merge pull request #15 from CS2103JAN2018-T09-B1/Week12Testing

Week12 testing update

kengsengg committed:
Merge branch 'Week12Testing' into wk12

kengsengg committed:
Merge pull request #189 from kengsengg/wk12

Added appointmentList that is saved locally and shown on UI

kengsengg committed:
Merge pull request #20 from CS2103JAN2018-T09-B1/Week12Testing

Week12 testing

TeyXinHui committed:
created a new score attribute to the student(l1r5, l1b4a, l1b4b, l1b4c, l1b4d)

TeyXinHui committed:
fix conflicts

TeyXinHui committed:
fix checkstyle

TeyXinHui committed:
fix checkstyle?

TeyXinHui committed:
checkstyle why

TeyXinHui committed:
Merge pull request #190 from TeyXinHui/master

scores attribute and fix conflicts

TeyXinHui committed:
print nok details method

TeyXinHui committed:
checkstyle

TeyXinHui committed:
Merge pull request #191 from TeyXinHui/master

print nok details method

johnnychanjx committed:
Merge pull request #21 from CS2103JAN2018-T09-B1/Week12Testing

Week12 testing

johnnychanjx committed:
Added UI support for phone, l1r5,cca

kengsengg committed:
Updated UGDG, About Us and PPP

kengsengg committed:
Updated picture file name

kengsengg committed:
Added whitespace for messages and adjusted settings for panel

johnnychanjx committed:
Updated UGDG

johnnychanjx committed:
Shows - when no L1R5 is out

kengsengg committed:
Merge pull request #192 from kengsengg/wk12testing

v1.5rc Enhancements

johnnychanjx committed:
Fixed Checkstyle

johnnychanjx committed:
Merge pull request #194 from johnnychanjx/Week13

Week13 Finalise UI Changes

johnnychanjx committed:
Use File Seperator to support multiple OSes

johnnychanjx committed:
Merge pull request #195 from johnnychanjx/Week13

Use File Seperator to support multiple OSes

kengsengg committed:
Merge pull request #21 from CS2103JAN2018-T09-B1/master

wk12 master

kengsengg committed:
Merge branch 'wk12testing' into Week12Testing

kengsengg committed:
Merge pull request #22 from CS2103JAN2018-T09-B1/Week12Testing

Week12 tests

kengsengg committed:
Merge pull request #196 from kengsengg/wk12testing

Updated v1.5rc

chuakunhong committed:
Need to keep this first

TeyXinHui committed:
change colour theme for ui

TeyXinHui committed:
Merge pull request #197 from TeyXinHui/master

change colour theme for ui

TeyXinHui committed:
Merge pull request #16 from CS2103JAN2018-T09-B1/master

update for 1.5

TeyXinHui committed:
remove score attribute

TeyXinHui committed:
checkstyle

TeyXinHui committed:
missing import?

TeyXinHui committed:
checkstyle

TeyXinHui committed:
Merge pull request #198 from TeyXinHui/master

remove scores attribute

chuakunhong committed:
Pass test and checkstyle!

chuakunhong committed:
Merge branch 'Week12' of https://github.com/chuakunhong/main into Week12

# Conflicts:
#	src/test/java/seedu/address/logic/commands/StreamCommandTest.java

chuakunhong committed:
Merge pull request #23 from CS2103JAN2018-T09-B1/master

Updating again.

chuakunhong committed:
Merge branch 'master' into Week12

# Conflicts:
#	src/main/java/seedu/address/logic/commands/AddCcaCommand.java
#	src/main/java/seedu/address/logic/commands/AddCommand.java
#	src/main/java/seedu/address/logic/commands/AddInjuriesHistoryCommand.java
#	src/main/java/seedu/address/logic/commands/AddRemarkCommand.java
#	src/main/java/seedu/address/logic/commands/DeleteCcaCommand.java
#	src/main/java/seedu/address/logic/commands/EditCommand.java
#	src/main/java/seedu/address/model/AddressBook.java
#	src/main/java/seedu/address/model/util/SampleDataUtil.java
#	src/main/resources/StudentPage/Alice Pauline.html
#	src/main/resources/StudentPage/Xinhui.html
#	src/test/java/seedu/address/storage/XmlAdaptedPersonTest.java

chuakunhong committed:
Merge pull request #199 from chuakunhong/Week12

Week12

johnnychanjx committed:
Merge pull request #22 from CS2103JAN2018-T09-B1/Week12Testing

Week12 testing

TeyXinHui committed:
Merge pull request #17 from CS2103JAN2018-T09-B1/master

update week 12

TeyXinHui committed:
edit invalid messages

TeyXinHui committed:
add new editSubjectCommand

TeyXinHui committed:
add editCommandTest

TeyXinHui committed:
add EditSubjectCommandTest

TeyXinHui committed:
Merge pull request #201 from TeyXinHui/master

add new EditSubjectCommand and some tests

johnnychanjx committed:
Merge pull request #23 from CS2103JAN2018-T09-B1/master

mergy

kengsengg committed:
Edited name for AppointmentCommand and fixed checkstyle

kengsengg committed:
Merge branch 'wk12testing'

johnnychanjx committed:
Added new Command to change Theme

johnnychanjx committed:
Update UGDG

johnnychanjx committed:
Add support for addnok command

johnnychanjx committed:
Added UI support after undo command is called

johnnychanjx committed:
Add newline to css files

johnnychanjx committed:
Merge pull request #203 from johnnychanjx/Week13

Added theme command to change colour scheme of EduBuddy

johnnychanjx committed:
Update UI

johnnychanjx committed:
Change Gender: to Relatationship:

johnnychanjx committed:
Delete Ui.png

johnnychanjx committed:
fixing merging problem

johnnychanjx committed:
Merge pull request #204 from johnnychanjx/Week13

Week13

johnnychanjx committed:
fixing merging problem pt 2

johnnychanjx committed:
Merge pull request #205 from johnnychanjx/Week13

fixing merging problem pt 2

TeyXinHui committed:
Merge pull request #18 from CS2103JAN2018-T09-B1/master

update for week 12

chuakunhong committed:
Make remarks great again.

chuakunhong committed:
Merge pull request #24 from CS2103JAN2018-T09-B1/master

Updating my master branch.

TeyXinHui committed:
enhance parseSubjects to prevent the adding of duplicate subjects

chuakunhong committed:
Updated the phone and email.

TeyXinHui committed:
Merge pull request #207 from TeyXinHui/master

Fix adding of duplicate subject problem

TeyXinHui committed:
add new method to return unmodifiable Set<subject>

TeyXinHui committed:
add new addSubjectCommand

TeyXinHui committed:
add test cases for AddSubjectCommand

TeyXinHui committed:
checkstyle

TeyXinHui committed:
fix checkstyle

TeyXinHui committed:
checkstyle

TeyXinHui committed:
Merge pull request #209 from TeyXinHui/master

new addSubjectCommand

TeyXinHui committed:
fix unable to add subject when there is duplicate subject bug

TeyXinHui committed:
Merge pull request #211 from TeyXinHui/master

fix unable to add subject when there is duplicate subject bug

kengsengg committed:
Implemented DeleteAppointmentCommand

kengsengg committed:
Fixed header comments and failed tests

chuakunhong committed:
pass check style!

chuakunhong committed:
Merge pull request #212 from chuakunhong/Week12

Week12

chuakunhong committed:
Updated addremarkcommand test.

chuakunhong committed:
Done with the addremark comman parser test!

TeyXinHui committed:
update userguide

TeyXinHui committed:
update project portfolio

TeyXinHui committed:
Merge pull request #213 from TeyXinHui/master

Update userguide and ppp

chuakunhong committed:
DeleteRemarkCommand Test done!

TeyXinHui committed:
Merge pull request #19 from CS2103JAN2018-T09-B1/master

update week 12

TeyXinHui committed:
add more pictures/update pictures in ugdg

TeyXinHui committed:
Merge pull request #215 from TeyXinHui/master

update more pictures for ugdg

kengsengg committed:
Implemented restrictions for adding of appointments

TeyXinHui committed:
update ugdg

TeyXinHui committed:
Merge pull request #216 from TeyXinHui/master

update ugdg and ppp

TeyXinHui committed:
update ug and pp again

TeyXinHui committed:
create tests for nextOfKinCommand

TeyXinHui committed:
Merge pull request #217 from TeyXinHui/master

update ug and pp again

kengsengg committed:
Added the name attribute to appointment

TeyXinHui committed:
create tests for nokcommandparser

TeyXinHui committed:
Merge pull request #218 from TeyXinHui/master

create tests for nokcommandparser

TeyXinHui committed:
add new tests cases for addccacommandparser

TeyXinHui committed:
checkstyle

TeyXinHui committed:
Merge pull request #219 from TeyXinHui/master

add new tests cases for addccacommandparser and addccacommand

chuakunhong committed:
Pass test for deleteremarks command and parser and addinjuries parser.

kengsengg committed:
Able to add and delete event on calendar based on given ID

chuakunhong committed:
Updated the test for injuries!

chuakunhong committed:
Pass checkstyle!

chuakunhong committed:
Merge branch 'master' into Week12

chuakunhong committed:
Update CommandTestUtil.java

kengsengg committed:
Able to view calendar but unable to pass in user input

chuakunhong committed:
Merge pull request #220 from chuakunhong/Week12

Week12

kengsengg committed:
Merge branch 'wk13testing' into master

kengsengg committed:
Merge pull request #23 from CS2103JAN2018-T09-B1/master

wk13 tests

chuakunhong committed:
Merge pull request #25 from CS2103JAN2018-T09-B1/master

Updating my main

chuakunhong committed:
getting rid of delete cca

chuakunhong committed:
Merge branch 'Week12'

kengsengg committed:
Merge branch 'wk13testing' into wk13

kengsengg committed:
Merge pull request #24 from kengsengg/wk13

Wk13

chuakunhong committed:
Change to make the command look better.

kengsengg committed:
Added AddAppointmentCommandTest class, collated files and fixed messages

kengsengg committed:
fixed eof error

kengsengg committed:
Merge pull request #221 from kengsengg/wk13testing

Added DeleteApptCommand and ViewCommand

johnnychanjx committed:
Merge pull request #24 from CS2103JAN2018-T09-B1/master

pr to m3

johnnychanjx committed:
Add Change Theme Tests

johnnychanjx committed:
Merge pull request #222 from johnnychanjx/Week13

Add Change theme tests

TeyXinHui committed:
Merge pull request #20 from CS2103JAN2018-T09-B1/master

week 12

chuakunhong committed:
Updated the user guide. Still missing some pictures inside.

kengsengg committed:
Implemented ViewCommand with working Google Calendar page

chuakunhong committed:
Updated the dg.

chuakunhong committed:
Merge branch 'master' into master

TeyXinHui committed:
update ug

chuakunhong committed:
Update AddressBookParser.java

chuakunhong committed:
Update AddressBookParser.java

TeyXinHui committed:
Merge pull request #224 from TeyXinHui/master

update ug and fix bug

chuakunhong committed:
Merge pull request #223 from chuakunhong/master

Updated the ugdg.

TeyXinHui committed:
Merge pull request #21 from CS2103JAN2018-T09-B1/master

update for v1.5

TeyXinHui committed:
update ug

johnnychanjx committed:
Updated PPP

TeyXinHui committed:
Merge branch 'master' of https://github.com/TeyXinHui/main

TeyXinHui committed:
solve adding of people witho duplicate ic

TeyXinHui committed:
solve test and prevent adding of duplicate ic from editcommand

TeyXinHui committed:
Merge pull request #225 from TeyXinHui/master

update ug

chuakunhong committed:
Updated the user guide part for adding and deleting of injuries.

kengsengg committed:
Updated UG and tests

TeyXinHui committed:
solves some test for editcommand

TeyXinHui committed:
checkstyle

kengsengg committed:
Fixed checkstyle

TeyXinHui committed:
Merge pull request #226 from TeyXinHui/master

solves some test for editcommand

TeyXinHui committed:
solves the failed tests

TeyXinHui committed:
Merge pull request #227 from TeyXinHui/master

solves the failed tests

chuakunhong committed:
Updated ugdg

chuakunhong committed:
Merge branch 'master' of https://github.com/chuakunhong/main

kengsengg committed:
Merge branch 'master' into wk13testing

kengsengg committed:
Merge pull request #228 from kengsengg/wk13testing

Updated UG and implemented ViewCommand

TeyXinHui committed:
update ugdg

TeyXinHui committed:
Merge pull request #229 from TeyXinHui/master

update ugdg

chuakunhong committed:
Merge pull request #230 from chuakunhong/master

Updated ugdg

chuakunhong committed:
Merge pull request #26 from CS2103JAN2018-T09-B1/master

Updating my main

chuakunhong committed:
Collated properly.

johnnychanjx committed:
Merge pull request #231 from johnnychanjx/Week13

Updated PPP

chuakunhong committed:
Updated UGDG

chuakunhong committed:
Merge pull request #232 from chuakunhong/master

Updated the UGDG

johnnychanjx committed:
Added collate stuff

johnnychanjx committed:
Merge pull request #233 from johnnychanjx/Week13

Added collate stuff

johnnychanjx committed:
collate stuff

johnnychanjx committed:
Merge pull request #234 from johnnychanjx/Week13

collate stuff

chuakunhong committed:
Merge pull request #27 from CS2103JAN2018-T09-B1/master

Updating my main.

chuakunhong committed:
For collation!

chuakunhong committed:
Merge pull request #235 from chuakunhong/master

For collation

johnnychanjx committed:
updated about us

johnnychanjx committed:
Merge pull request #236 from johnnychanjx/Week13

updated about us

johnnychanjx committed:
collate stuff

johnnychanjx committed:
Updated user guide

johnnychanjx committed:
Merge branch 'master' into Week13

johnnychanjx committed:
Merge pull request #237 from johnnychanjx/Week13

Week13

kengsengg committed:
Updated UGDG, PP

kengsengg committed:
Resolved conflicts

kengsengg committed:
Merge pull request #25 from CS2103JAN2018-T09-B1/master

wk13

chuakunhong committed:
Updated.

chuakunhong committed:
Merge pull request #238 from chuakunhong/master

Updated.

kengsengg committed:
Merge pull request #26 from CS2103JAN2018-T09-B1/master

last

chuakunhong committed:
update again

chuakunhong committed:
Merge pull request #239 from chuakunhong/master

update again

kengsengg committed:
Merge pull request #27 from CS2103JAN2018-T09-B1/master

last

kengsengg committed:
Collated

kengsengg committed:
Merge pull request #240 from kengsengg/wk13testing

v1.5

chuakunhong committed:
Update again

chuakunhong committed:
Merge pull request #28 from CS2103JAN2018-T09-B1/master

Update

chuakunhong committed:
Merge pull request #241 from chuakunhong/master

Update again.

chuakunhong committed:
update again

chuakunhong committed:
Merge pull request #242 from chuakunhong/master

update again
